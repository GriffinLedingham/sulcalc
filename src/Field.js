import { Terrains, Weathers, maxGen } from "./utilities";

export default class Field {
  constructor(field = {}) {
    this.gen = Number(field.gen ?? maxGen);
    this.multiBattle = Boolean(field.multiBattle);
    this.invertedBattle = Boolean(field.invertedBattle);

    this.weather = Number(field.weather ?? Weathers.CLEAR);
    this.airLock = Boolean(field.airLock);

    this.waterSport = Boolean(field.waterSport);
    this.mudSport = Boolean(field.mudSport);

    this.magicRoom = Boolean(field.magicRoom);
    this.wonderRoom = Boolean(field.wonderRoom);
    this.gravity = Boolean(field.gravity);

    this.terrain = field.terrain ?? Terrains.NO_TERRAIN;

    this.fairyAura = Boolean(field.fairyAura);
    this.darkAura = Boolean(field.darkAura);
    this.auraBreak = Boolean(field.auraBreak);

    this.ionDeluge = Boolean(field.ionDeluge);
  }

  effectiveWeather() {
    return this.airLock ? Weathers.CLEAR : this.weather;
  }

  isClearWeather() {
    return this.effectiveWeather() === Weathers.CLEAR;
  }

  rain() {
    return [Weathers.RAIN, Weathers.HEAVY_RAIN].includes(
      this.effectiveWeather()
    );
  }

  sun() {
    return [Weathers.SUN, Weathers.HARSH_SUN].includes(this.effectiveWeather());
  }

  sand() {
    return this.effectiveWeather() === Weathers.SAND;
  }

  hail() {
    return this.effectiveWeather() === Weathers.HAIL;
  }

  heavyRain() {
    return this.effectiveWeather() === Weathers.HEAVY_RAIN;
  }

  harshSun() {
    return this.effectiveWeather() === Weathers.HARSH_SUN;
  }

  strongWinds() {
    return this.effectiveWeather() === Weathers.STRONG_WINDS;
  }

  grassyTerrain() {
    return this.terrain === Terrains.GRASSY_TERRAIN;
  }

  mistyTerrain() {
    return this.terrain === Terrains.MISTY_TERRAIN;
  }

  electricTerrain() {
    return this.terrain === Terrains.ELECTRIC_TERRAIN;
  }

  psychicTerrain() {
    return this.terrain === Terrains.PSYCHIC_TERRAIN;
  }
}
