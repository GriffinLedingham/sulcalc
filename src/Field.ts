import { Generation, Terrain, Weather, maxGen } from "./utilities";

export type FieldOptions = Partial<Field>;

export default class Field {
  public gen: Generation = maxGen;
  public weather: Weather = Weather.CLEAR;
  public terrain: Terrain = Terrain.NO_TERRAIN;
  public multiBattle: boolean = false;
  public invertedBattle: boolean = false;
  public airLock: boolean = false;
  public waterSport: boolean = false;
  public mudSport: boolean = false;
  public magicRoom: boolean = false;
  public wonderRoom: boolean = false;
  public gravity: boolean = false;
  public fairyAura: boolean = false;
  public darkAura: boolean = false;
  public auraBreak: boolean = false;
  public ionDeluge: boolean = false;

  public constructor(field: Field | FieldOptions = {}) {
    Object.assign(this, field);
  }

  public effectiveWeather(): Weather {
    return this.airLock ? Weather.CLEAR : this.weather;
  }

  public isClearWeather(): boolean {
    return this.effectiveWeather() === Weather.CLEAR;
  }

  public rain(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.RAIN || weather === Weather.HEAVY_RAIN;
  }

  public sun(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.SUN || weather === Weather.HARSH_SUN;
  }

  public sand(): boolean {
    return this.effectiveWeather() === Weather.SAND;
  }

  public hail(): boolean {
    return this.effectiveWeather() === Weather.HAIL;
  }

  public heavyRain(): boolean {
    return this.effectiveWeather() === Weather.HEAVY_RAIN;
  }

  public harshSun(): boolean {
    return this.effectiveWeather() === Weather.HARSH_SUN;
  }

  public strongWinds(): boolean {
    return this.effectiveWeather() === Weather.STRONG_WINDS;
  }

  public grassyTerrain(): boolean {
    return this.terrain === Terrain.GRASSY_TERRAIN;
  }

  public mistyTerrain(): boolean {
    return this.terrain === Terrain.MISTY_TERRAIN;
  }

  public electricTerrain(): boolean {
    return this.terrain === Terrain.ELECTRIC_TERRAIN;
  }

  public psychicTerrain(): boolean {
    return this.terrain === Terrain.PSYCHIC_TERRAIN;
  }
}
