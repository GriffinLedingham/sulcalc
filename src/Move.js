import { clamp, defaultTo } from "lodash";
import {
  Gens,
  DamageClasses,
  Stats,
  Types,
  Weathers,
  maxGen
} from "./utilities";
import {
  moveId,
  moveName,
  movePower,
  moveDamageClass,
  moveType,
  priority,
  requiresRecharge,
  moveIgnoresAbilities,
  recoil,
  hasPunchFlag,
  hasContactFlag,
  hasSoundFlag,
  hasPowderFlag,
  hasBiteFlag,
  hasPulseFlag,
  hasBulletFlag,
  hasSecondaryEffect,
  minHits,
  maxHits,
  hitsMultipleTargets,
  zMovePower,
  isOhkoMove
} from "./info";

const zMoves = {
  [Types.NORMAL]: "Breakneck Blitz",
  [Types.FIGHTING]: "All-Out Pummeling",
  [Types.FLYING]: "Supersonic Skystrike",
  [Types.POISON]: "Acid Downpour",
  [Types.GROUND]: "Tectonic Rage",
  [Types.ROCK]: "Continental Crush",
  [Types.BUG]: "Savage Spin-Out",
  [Types.GHOST]: "Never-Ending Nightmare",
  [Types.STEEL]: "Corkscrew Crash",
  [Types.FIRE]: "Inferno Overdrive",
  [Types.WATER]: "Hydro Vortex",
  [Types.GRASS]: "Bloom Doom",
  [Types.ELECTRIC]: "Gigavolt Havoc",
  [Types.PSYCHIC]: "Shattered Psyche",
  [Types.ICE]: "Subzero Slammer",
  [Types.DRAGON]: "Devastating Drake",
  [Types.DARK]: "Black Hole Eclipse",
  [Types.FAIRY]: "Twinkle Tackle"
};

const hiddenPowerStatOrder = [
  Stats.SPD,
  Stats.SATK,
  Stats.HP,
  Stats.DEF,
  Stats.SDEF,
  Stats.ATK
];

function ivsCmp(a, b) {
  for (const stat of hiddenPowerStatOrder) {
    if (a[stat] !== b[stat]) return b[stat] - a[stat];
  }
  /* istanbul ignore next */
  return 0;
}

export default class Move {
  constructor(move = {}) {
    this.gen = defaultTo(Number(move.gen), maxGen);
    this.id = moveId(move.id);
    if (typeof move.name === "string") this.name = move.name;

    this.user = move.user;
    this.target = move.target;

    if (move.numberOfHits == null) {
      if (this.user && this.user.ability.name === "Skill Link") {
        this.numberOfHits = maxHits(this.id, this.gen);
      } else if (
        this.name === "Water Shuriken" &&
        this.user &&
        this.user.name === "Greninja-Ash"
      ) {
        this.numberOfHits = 3;
      } else {
        this.numberOfHits = 0;
      }
    } else {
      this.numberOfHits = Number(move.numberOfHits);
    }

    this.critical = Boolean(move.critical);
    this.zMove = Boolean(move.zMove);
    this.beatUpHit = defaultTo(Number(move.beatUpHit), 0);
    this.secondHit = Boolean(move.secondHit);
    this.meFirst = Boolean(move.meFirst);
    this.furyCutter = defaultTo(Number(move.furyCutter), 0);
    this.echoedVoice = defaultTo(Number(move.echoedVoice), 0);
    this.trumpPP = defaultTo(Number(move.trumpPP), 5);
    this.roundBoost = Boolean(move.roundBoost);
    this.minimize = Boolean(move.minimize);
    this.dig = Boolean(move.dig);
    this.dive = Boolean(move.dive);
    this.fly = Boolean(move.fly);
    this.present = defaultTo(Number(move.present), -1);
  }

  get name() {
    if (this.zMove) {
      if (this.user && this.id === this.user.item.zMoveTransformsFrom()) {
        return moveName(this.user.item.zMoveTransformsTo());
      }
      return zMoves[this.type()];
    }
    return moveName(this.id);
  }

  set name(moveName) {
    this.id = moveId(String(moveName));
  }

  power() {
    if (this.zMove) {
      if (this.user && this.id === this.user.item.zMoveTransformsFrom()) {
        return movePower(this.user.item.zMoveTransformsTo(), this.gen);
      }
      return zMovePower(this.id, this.gen);
    }
    return movePower(this.id, this.gen);
  }

  type() {
    return moveType(this.id, this.gen);
  }

  damageClass() {
    return moveDamageClass(this.id, this.gen);
  }

  isPhysical() {
    return this.damageClass() === DamageClasses.PHYSICAL;
  }

  isSpecial() {
    return this.damageClass() === DamageClasses.SPECIAL;
  }

  isOther() {
    return this.damageClass() === DamageClasses.OTHER;
  }

  priority() {
    return priority(this.id, this.gen);
  }

  isPsyshockLike() {
    return ["Psyshock", "Psystrike", "Secret Sword"].includes(this.name);
  }

  hasRecoil() {
    return Boolean(recoil(this.id, this.gen));
  }

  affectedBySheerForce() {
    return hasSecondaryEffect(this.id, this.gen);
  }

  isPunch() {
    return hasPunchFlag(this.id, this.gen);
  }

  isContact() {
    return hasContactFlag(this.id, this.gen);
  }

  isSound() {
    return hasSoundFlag(this.id, this.gen);
  }

  isPowder() {
    return hasPowderFlag(this.id, this.gen);
  }

  isBite() {
    return hasBiteFlag(this.id, this.gen);
  }

  isPulse() {
    return hasPulseFlag(this.id, this.gen);
  }

  isBall() {
    return hasBulletFlag(this.id, this.gen);
  }

  minHits() {
    return minHits(this.id, this.gen);
  }

  maxHits() {
    return maxHits(this.id, this.gen);
  }

  hitsMultipleTimes() {
    return this.maxHits() > 1;
  }

  hasMultipleTargets() {
    return hitsMultipleTargets(this.id, this.gen);
  }

  isOhko() {
    return isOhkoMove(this.id, this.gen);
  }

  isExplosion() {
    return ["Explosion", "Mind Blown", "Self-Destruct"].includes(this.name);
  }

  requiresRecharge() {
    return requiresRecharge(this.id, this.gen);
  }

  isRecklessBoosted() {
    return (
      this.hasRecoil() ||
      this.name === "Jump Kick" ||
      this.name === "High Jump Kick"
    );
  }

  canCrit() {
    if (this.gen >= Gens.HGSS) return true;
    return ![
      "Reversal",
      "Flail",
      "Future Sight",
      "Doom Desire",
      "Spit Up"
    ].includes(this.name);
  }

  affectedByParentalBond() {
    return ![
      "Endeavor",
      "Explosion",
      "Final Gambit",
      "Fling",
      "Self-Destruct"
    ].includes(this.name);
  }

  optimalHappiness() {
    return this.name === "Return" ? 255 : 0;
  }

  usesHappiness() {
    return this.name === "Return" || this.name === "Frustration";
  }

  ignoresAbilities() {
    return moveIgnoresAbilities(this.id, this.gen);
  }

  ignoresDefenseBoosts() {
    return ["Chip Away", "Sacred Sword", "Darkest Lariat"].includes(this.name);
  }

  boostedByDig() {
    return this.name === "Earthquake" || this.name === "Magnitude";
  }

  boostedByDive() {
    return this.name === "Surf" || this.name === "Whirlpool";
  }

  boostedByFly() {
    return this.name === "Gust" || this.name === "Twister";
  }

  boostedByMinimize() {
    switch (this.name) {
      case "Stomp":
        return this.gen >= Gens.GSC;
      case "Astonish":
      case "Extrasensory":
      case "Needle Arm":
        return this.gen === Gens.ADV;
      case "Steamroller":
        return this.gen >= Gens.B2W2;
      case "Body Slam":
      case "Dragon Rush":
      case "Flying Press":
      case "Shadow Force":
      case "Phantom Force":
        return this.gen >= Gens.ORAS;
      case "Heavy Slam":
        return this.gen >= Gens.SM;
      default:
        return false;
    }
  }

  weakenedByGrassyTerrain() {
    return ["Bulldoze", "Earthquake", "Magnitude"].includes(this.name);
  }

  usesMaxAttackingStat() {
    return ["Photon Geyser", "Light That Burns the Sky"].includes(this.name);
  }

  isHiddenPower() {
    return this.name.startsWith("Hidden Power");
  }

  static hiddenPowers(typeId, gen) {
    if (gen < Gens.ADV) {
      if (typeId < 1 || typeId > 16) {
        return [];
      }
      const t = typeId - 1;
      return [
        [
          ((t & 4) << 1) | ((t & 1) << 2) | 3,
          (t >> 2) | 12,
          (t & 3) | 12,
          15,
          15,
          15
        ]
      ];
    }

    if (gen < Gens.SM) {
      const hiddenPowers = [];
      // check if each possible hidden power matches the type
      for (let i = 0; i < 64; i++) {
        const ivs = [
          i & 1 ? 31 : 30,
          i & 2 ? 31 : 30,
          i & 4 ? 31 : 30,
          i & 8 ? 31 : 30,
          i & 16 ? 31 : 30,
          i & 32 ? 31 : 30
        ];
        if (typeId === Move.hiddenPowerType(ivs)) {
          hiddenPowers.push(ivs);
        }
      }
      return hiddenPowers.sort(ivsCmp);
    }

    return [[31, 31, 31, 31, 31, 31]];
  }

  static hiddenPowerBp(ivs, gen) {
    // differs for gsc
    if (gen < Gens.ADV) {
      const bits =
        (ivs[Stats.SPC] >> 3) |
        ((ivs[Stats.SPD] >> 2) & 2) |
        ((ivs[Stats.DEF] >> 1) & 4) |
        (ivs[Stats.ATK] & 8);
      return 31 + Math.trunc((5 * bits + (ivs[Stats.SPC] & 3)) / 2);
    }

    // just a weird formula involving the second bit of the pokemon's IVs
    if (gen < Gens.ORAS) {
      const bits =
        ((ivs[Stats.HP] & 2) >> 1) |
        (ivs[Stats.ATK] & 2) |
        ((ivs[Stats.DEF] & 2) << 1) |
        ((ivs[Stats.SPD] & 2) << 2) |
        ((ivs[Stats.SATK] & 2) << 3) |
        ((ivs[Stats.SDEF] & 2) << 4);
      return 30 + Math.trunc((bits * 40) / 63);
    }

    // constant in oras
    return 60;
  }

  static hiddenPowerType(ivs, gen) {
    // Another weird formula involving the first bit (even/odd) of the
    // pokemon's IVs. Also differs in gsc.
    if (gen < Gens.ADV) {
      return 1 + (((ivs[Stats.ATK] & 3) << 2) | (ivs[Stats.DEF] & 3));
    }

    const bits =
      (ivs[Stats.HP] & 1) |
      ((ivs[Stats.ATK] & 1) << 1) |
      ((ivs[Stats.DEF] & 1) << 2) |
      ((ivs[Stats.SPD] & 1) << 3) |
      ((ivs[Stats.SATK] & 1) << 4) |
      ((ivs[Stats.SDEF] & 1) << 5);
    return 1 + Math.trunc((bits * 15) / 63);
  }

  static flail(currentHealth, totalHealth, gen) {
    if (gen === Gens.HGSS) {
      const p = Math.trunc((64 * currentHealth) / totalHealth);
      if (p < 2) return 200;
      if (p < 6) return 150;
      if (p < 13) return 100;
      if (p < 22) return 80;
      if (p < 43) return 40;
      return 20;
    }

    const p = Math.trunc((48 * currentHealth) / totalHealth);
    if (p <= 1) return 200;
    if (p <= 4) return 150;
    if (p <= 9) return 100;
    if (p <= 16) return 80;
    if (p <= 32) return 40;
    return 20;
  }

  static magnitude(magnitude) {
    return [10, 30, 50, 70, 90, 110, 150][magnitude - 4];
  }

  static weatherBall(weather) {
    if (weather === Weathers.SUN || weather === Weathers.HARSH_SUN) {
      return Types.FIRE;
    }
    if (weather === Weathers.RAIN || weather === Weathers.HEAVY_RAIN) {
      return Types.WATER;
    }
    if (weather === Weathers.SAND) {
      return Types.ROCK;
    }
    if (weather === Weathers.HAIL) {
      return Types.ICE;
    }
    return Types.NORMAL;
  }

  static trumpCard(ppLeft) {
    return [200, 80, 60, 50, 40][clamp(ppLeft, 0, 4)];
  }

  static electroBall(attackerSpeed, defenderSpeed) {
    const s = attackerSpeed / Math.max(1, defenderSpeed);
    if (s >= 4) return 150;
    if (s >= 3) return 120;
    if (s >= 2) return 80;
    if (s >= 1) return 60;
    return 40;
  }

  static gyroBall(attackerSpeed, defenderSpeed) {
    return Math.min(
      150,
      Math.trunc((25 * defenderSpeed) / Math.max(attackerSpeed, 1))
    );
  }

  static grassKnot(weight) {
    // w in kg * 10
    if (weight >= 2000) return 120;
    if (weight >= 1000) return 100;
    if (weight >= 500) return 80;
    if (weight >= 250) return 60;
    if (weight >= 100) return 40;
    return 20;
  }

  static heavySlam(attackerWeight, defenderWeight) {
    const w = Math.trunc(attackerWeight / defenderWeight);
    if (w >= 5) return 120;
    if (w === 4) return 100;
    if (w === 3) return 80;
    if (w === 2) return 60;
    return 40;
  }

  static punishment(boosts) {
    let statBoostTotal = 0;
    for (const boost of boosts.slice(1, 8)) {
      if (boost > 0) {
        statBoostTotal += boost;
      }
    }
    return Math.min(120, 60 + 20 * statBoostTotal);
  }

  static storedPower(boosts) {
    let statBoostTotal = 0;
    for (const boost of boosts.slice(1, 8)) {
      if (boost > 0) {
        statBoostTotal += boost;
      }
    }
    return 20 + 20 * statBoostTotal;
  }

  static frustration(happiness) {
    return Math.max(1, Math.trunc(((255 - happiness) * 10) / 25));
  }

  static return(happiness) {
    return Math.max(1, Math.trunc((happiness * 10) / 25));
  }

  static eruption(hp, totalHp) {
    return Math.max(1, Math.trunc((150 * hp) / totalHp));
  }
}
