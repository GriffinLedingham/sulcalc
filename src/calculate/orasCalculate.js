import {
  Stats,
  Types,
  damageVariation,
  applyMod,
  chainMod
} from "../utilities";
import moveInfo from "./moveInfo";

export default (attacker, defender, move, field) => {
  const {
    moveType,
    movePower,
    effectiveness,
    superEffective,
    notVeryEffective,
    fail
  } = moveInfo(attacker, defender, move, field);
  if (fail || effectiveness[0] === 0) return [0];

  const defStat = field.wonderRoom ? Stats.SDEF : Stats.DEF;
  const sdefStat = field.wonderRoom ? Stats.DEF : Stats.SDEF;
  const unawareA = attacker.ability.name === "Unaware";
  const unawareD = defender.ability.name === "Unaware";
  let atk, def, satk, sdef;
  if (move.name === "Foul Play") {
    if (unawareA) {
      def = defender.stat(defStat);
      sdef = defender.stat(sdefStat);
      atk = defender.stat(Stats.ATK);
    } else if (move.critical) {
      def = Math.min(defender.stat(defStat), defender.boostedStat(defStat));
      sdef = Math.min(defender.stat(sdefStat), defender.boostedStat(sdefStat));
      atk = Math.max(defender.stat(Stats.ATK), defender.boostedStat(Stats.ATK));
    } else {
      def = defender.boostedStat(defStat);
      sdef = defender.boostedStat(sdefStat);
      atk = defender.boostedStat(Stats.ATK);
    }

    if (unawareD) {
      satk = attacker.stat(Stats.SATK);
    } else if (move.critical) {
      satk = Math.max(
        attacker.stat(Stats.SATK),
        attacker.boostedStat(Stats.SATK)
      );
    } else {
      satk = attacker.boostedStat(Stats.SATK);
    }
  } else if (move.ignoresDefenseBoosts()) {
    def = defender.stat(defStat);
    sdef = defender.stat(sdefStat);

    if (unawareD) {
      atk = attacker.stat(Stats.ATK);
      satk = attacker.stat(Stats.SATK);
    } else if (move.critical) {
      atk = Math.max(attacker.stat(Stats.ATK), attacker.boostedStat(Stats.ATK));
      satk = Math.max(
        attacker.stat(Stats.SATK),
        attacker.boostedStat(Stats.SATK)
      );
    } else {
      atk = attacker.boostedStat(Stats.ATK);
      satk = attacker.boostedStat(Stats.SATK);
    }
  } else {
    if (unawareA) {
      def = defender.stat(defStat);
      sdef = defender.stat(sdefStat);
    } else if (move.critical) {
      def = Math.min(defender.stat(defStat), defender.boostedStat(defStat));
      sdef = Math.min(defender.stat(sdefStat), defender.boostedStat(sdefStat));
    } else {
      def = defender.boostedStat(defStat);
      sdef = defender.boostedStat(sdefStat);
    }

    if (unawareD) {
      atk = attacker.stat(Stats.ATK);
      satk = attacker.stat(Stats.SATK);
    } else if (move.critical) {
      atk = Math.max(attacker.stat(Stats.ATK), attacker.boostedStat(Stats.ATK));
      satk = Math.max(
        attacker.stat(Stats.SATK),
        attacker.boostedStat(Stats.SATK)
      );
    } else {
      atk = attacker.boostedStat(Stats.ATK);
      satk = attacker.boostedStat(Stats.SATK);
    }
  }

  let atkMod = 0x1000;
  let satkMod = 0x1000;

  if (
    defender.ability.name === "Thick Fat" &&
    (moveType === Types.FIRE || moveType === Types.ICE)
  ) {
    atkMod = chainMod(0x800, atkMod);
    satkMod = chainMod(0x800, satkMod);
  }

  switch (attacker.ability.name) {
    case "Guts":
      if (defender.status) atkMod = chainMod(0x1800, atkMod);
      break;
    case "Plus":
      if (attacker.minus) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Minus":
      if (attacker.plus) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Defeatist":
      if (attacker.currentHp * 2 <= attacker.stat(Stats.HP)) {
        atkMod = chainMod(0x800, atkMod);
        satkMod = chainMod(0x800, satkMod);
      }
      break;
    case "Huge Power":
    case "Pure Power":
      atkMod = chainMod(0x2000, atkMod);
      break;
    case "Solar Power":
      if (field.sun()) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Hustle":
      atk = applyMod(0x1800, atk);
      break;
    case "Flash Fire":
      if (attacker.flashFire && moveType === Types.FIRE) {
        atkMod = chainMod(0x1800, atkMod);
        satkMod = chainMod(0x1800, satkMod);
      }
      break;
    case "Slow Start":
      if (attacker.slowStart) atkMod = chainMod(0x800, atkMod);
      break;
    default:
      if (attacker.pinchAbilityActivated(moveType)) {
        // blaze, torrent, overgrow, ...
        atkMod = chainMod(0x1800, atkMod);
        satkMod = chainMod(0x1800, satkMod);
      }
  }

  if (attacker.flowerGift && field.sun()) {
    atkMod = chainMod(0x1800, atkMod);
  }

  switch (attacker.item.name) {
    case "Deep Sea Tooth":
      if (attacker.name === "Clamperl") {
        satkMod = chainMod(0x2000, satkMod);
      }
      break;
    case "Soul Dew":
      if (attacker.name === "Latias" || attacker.name === "Latios") {
        satkMod = chainMod(0x1800, satkMod);
      }
      break;
    case "Choice Band":
      atkMod = chainMod(0x1800, atkMod);
      break;
    case "Choice Specs":
      satkMod = chainMod(0x1800, satkMod);
      break;
    default:
      if (attacker.thickClubBoosted()) {
        atkMod = chainMod(0x2000, atkMod);
      } else if (attacker.lightBallBoosted()) {
        atkMod = chainMod(0x2000, atkMod);
        satkMod = chainMod(0x2000, satkMod);
      }
  }

  atk = applyMod(atkMod, atk);
  satk = applyMod(satkMod, satk);

  if (field.sand() && defender.stab(Types.ROCK)) {
    sdef = applyMod(0x1800, sdef);
  }

  let defMod = 0x1000;
  let sdefMod = 0x1000;

  if (defender.ability.name === "Marvel Scale" && defender.status) {
    defMod = chainMod(0x1800, defMod);
  } else if (defender.ability.name === "Grass Pelt" && field.grassyTerrain()) {
    defMod = chainMod(0x1800, defMod);
  }

  if (defender.flowerGift && field.sun()) {
    sdefMod = chainMod(0x1800, sdefMod);
  }

  switch (defender.item.name) {
    case "Deep Sea Scale":
      if (defender.name === "Clamperl") {
        sdefMod = chainMod(0x1800, sdefMod);
      }
      break;
    case "Metal Powder":
      if (defender.name === "Ditto") {
        defMod = chainMod(0x2000, defMod);
      }
      break;
    case "Eviolite":
      if (defender.hasEvolution()) {
        defMod = chainMod(0x1800, defMod);
        sdefMod = chainMod(0x1800, sdefMod);
      }
      break;
    case "Soul Dew":
      if (defender.name === "Latias" || defender.name === "Latios") {
        sdefMod = chainMod(0x1800, sdefMod);
      }
      break;
    case "Assault Vest":
      sdefMod = chainMod(0x1800, sdefMod);
      break;
    /* no default */
  }

  def = applyMod(defMod, def);
  sdef = applyMod(sdefMod, sdef);

  let a, d;
  if (move.isPsyshockLike()) {
    a = satk;
    d = def;
  } else if (move.isPhysical()) {
    a = atk;
    d = def;
  } else if (move.isSpecial()) {
    a = satk;
    d = sdef;
  } else {
    return [0];
  }

  let baseDamage = Math.trunc(
    Math.trunc((Math.trunc((2 * attacker.level) / 5 + 2) * movePower * a) / d) /
      50
  );
  baseDamage += 2;

  if (field.multiBattle && move.hasMultipleTargets()) {
    baseDamage = applyMod(0xc00, baseDamage);
  }

  if (move.name !== "Weather Ball") {
    if (
      (field.harshSun() && moveType === Types.WATER) ||
      (field.heavyRain() && moveType === Types.FIRE)
    ) {
      return [0];
    }
    if (field.sun()) {
      if (moveType === Types.FIRE) {
        baseDamage = applyMod(0x1800, baseDamage);
      } else if (moveType === Types.WATER) {
        baseDamage = applyMod(0x800, baseDamage);
      }
    } else if (field.rain()) {
      if (moveType === Types.WATER) {
        baseDamage = applyMod(0x1800, baseDamage);
      } else if (moveType === Types.FIRE) {
        baseDamage = applyMod(0x800, baseDamage);
      }
    }
  }

  if (
    ((field.grassyTerrain() && moveType === Types.GRASS) ||
      (field.electricTerrain() && moveType === Types.ELECTRIC) ||
      (field.psychicTerrain() && moveType === Types.PSYCHIC)) &&
    attacker.isGrounded(field)
  ) {
    baseDamage = applyMod(0x1800, baseDamage);
  }

  if (
    field.mistyTerrain() &&
    defender.isGrounded(field) &&
    moveType === Types.DRAGON
  ) {
    baseDamage = applyMod(0x800, baseDamage);
  }

  if (move.critical) {
    baseDamage = applyMod(0x1800, baseDamage);
  }

  let damages = damageVariation(baseDamage, 85, 100);

  if (attacker.stab(moveType) || attacker.ability.name === "Protean") {
    if (attacker.ability.name === "Adaptability") {
      damages = applyMod(0x2000, damages);
    } else {
      damages = applyMod(0x1800, damages);
    }
  }

  damages = damages.map(d =>
    Math.trunc((d * effectiveness[0]) / effectiveness[1])
  );

  if (
    attacker.isBurned() &&
    move.isPhysical() &&
    attacker.ability.name !== "Guts" &&
    move.name !== "Facade"
  ) {
    damages = damages.map(d => Math.trunc(d / 2));
  }

  damages = damages.map(d => Math.max(1, d));

  let finalMod = 0x1000;

  if (!move.critical && attacker.ability.name !== "Infiltrator") {
    if (defender.reflect && (move.isPhysical() || move.isPsyshockLike())) {
      finalMod = chainMod(field.multiBattle ? 0xa8f : 0x800, finalMod);
    }
    if (defender.lightScreen && move.isSpecial() && !move.isPsyshockLike()) {
      finalMod = chainMod(field.multiBattle ? 0xa8f : 0x800, finalMod);
    }
  }

  if (defender.multiscaleIsActive()) {
    finalMod = chainMod(0x800, finalMod);
  }

  if (
    defender.ability.name === "Fur Coat" &&
    (move.isPhysical() || move.isPsyshockLike())
  ) {
    finalMod = chainMod(0x800, finalMod);
  }

  if (notVeryEffective && attacker.ability.name === "Tinted Lens") {
    finalMod = chainMod(0x2000, finalMod);
  }

  if (defender.friendGuard) {
    finalMod = chainMod(0xc00, finalMod);
  }

  if (attacker.ability.name === "Sniper" && move.critical) {
    finalMod = chainMod(0x1800, finalMod);
  }

  if (superEffective && defender.ability.reducesSuperEffective()) {
    finalMod = chainMod(0xc00, finalMod);
  }

  if (field.grassyTerrain() && move.weakenedByGrassyTerrain()) {
    finalMod = chainMod(0x800, finalMod);
  }

  switch (attacker.item.name) {
    case "Metronome":
      if (attacker.metronome <= 4) {
        const mod = 0x1000 + attacker.metronome * 0x333;
        finalMod = chainMod(mod, finalMod);
      } else {
        finalMod = chainMod(0x2000, finalMod);
      }
      break;
    case "Expert Belt":
      if (superEffective) {
        finalMod = chainMod(0x1333, finalMod);
      }
      break;
    case "Life Orb":
      finalMod = chainMod(0x14cc, finalMod);
      break;
    /* no default */
  }

  if (
    moveType === defender.item.berryTypeResist() &&
    (superEffective || moveType === Types.NORMAL)
  ) {
    finalMod = chainMod(0x800, finalMod);
    defender.item.used = true;
  }

  if (
    (move.dig && move.boostedByDig()) ||
    (move.dive && move.boostedByDive()) ||
    (move.minimize && move.boostedByMinimize())
  ) {
    finalMod = chainMod(0x2000, finalMod);
  }

  damages = applyMod(finalMod, damages);

  return damages;
};
