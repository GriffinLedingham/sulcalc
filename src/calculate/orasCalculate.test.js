import orasCalculate from "sulcalc/calculate/orasCalculate";
import Pokemon from "sulcalc/Pokemon";
import Move from "sulcalc/Move";
import Field from "sulcalc/Field";
import { Gens, Natures } from "sulcalc/utilities";

const gen = Gens.ORAS;

describe("orasCalculate()", () => {
  let field;
  beforeEach(() => {
    field = new Field({ gen });
  });

  test("sanity check", () => {
    const attacker = new Pokemon({
      name: "Heatran",
      evs: [252, 0, 0, 0, 4, 252],
      nature: Natures.TIMID,
      gen
    });
    const defender = new Pokemon({
      name: "Landorus-Therian",
      evs: [252, 0, 216, 0, 24, 16],
      nature: Natures.IMPISH,
      gen
    });
    const move = new Move({
      name: "Lava Plume",
      gen
    });
    const damage = orasCalculate(attacker, defender, move, field);
    expect(Math.max(...damage)).toEqual(150);
  });

  test("Fairy Aura is a move power mod", () => {
    const xerneas = new Pokemon({
      name: "Xerneas",
      item: "Life Orb",
      nature: Natures.MODEST,
      evs: [0, 0, 0, 252, 0, 0],
      gen
    });
    const genesect = new Pokemon({
      name: "Genesect",
      nature: Natures.NAIVE,
      gen
    });
    const moonblast = new Move({ name: "Moonblast", gen });
    const auraField = new Field({ fairyAura: true, gen });
    expect(
      orasCalculate(xerneas, genesect, moonblast, auraField)
    ).toMatchSnapshot();
  });
});
