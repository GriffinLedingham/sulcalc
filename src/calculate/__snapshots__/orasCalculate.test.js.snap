// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orasCalculate() Fairy Aura is a move power mod 1`] = `
Array [
  172,
  173,
  175,
  178,
  179,
  182,
  183,
  186,
  187,
  190,
  191,
  194,
  195,
  198,
  199,
  203,
]
`;
